conans/model/graph_lock.py

```diff
@@ -0,0 +1,257 @@
+
+    def __init__(self, graph=None):
+        self._nodes = {}  # {numeric id: PREF or None}
+        self._edges = {}  # {numeric_id: [numeric_ids]}
+        if graph:
+            for node in graph.nodes:
+                if node.recipe == RECIPE_VIRTUAL:
+                    continue
+                dependencies = []
+                for edge in node.dependencies:
+                    dependencies.append(edge.dst.id)
+                python_reqs = getattr(node.conanfile, "python_requires", {})
+                python_reqs = [r.ref for _, r in python_reqs.items()] if python_reqs else None
+                graph_node = GraphLockNode(node.pref if node.ref else None, python_reqs,
+                                           node.conanfile.options.values, False)
+                self._nodes[node.id] = graph_node
+                self._edges[node.id] = dependencies
+
+    def root_node(self):
+        total = []
+        for list_ids in self._edges.values():
+            total.extend(list_ids)
+        roots = set(self._edges).difference(total)
+        assert len(roots) == 1
+        return self._nodes[roots.pop()]
+
+    @staticmethod
+    def from_dict(graph_json):
+        graph_lock = GraphLock()
+        for id_, node in graph_json["nodes"].items():
+            graph_lock._nodes[id_] = GraphLockNode.from_dict(node)
+        for id_, dependencies in graph_json["edges"].items():
+            graph_lock._edges[id_] = dependencies
+
+        return graph_lock
+
+    def as_dict(self):
+        result = {}
+        nodes = {}
+        for id_, node in self._nodes.items():
+            nodes[id_] = node.as_dict()
+        result["nodes"] = nodes
+        result["edges"] = self._edges
+        return result
+
+    def update_lock(self, new_lock):
+        for id_, node in new_lock._nodes.items():
+            if node.modified:
+                old_node = self._nodes[id_]
+                if old_node.modified:
+                    raise ConanException("Lockfile had already modified %s" % str(node.pref))
+                self._nodes[id_] = node
+
+    def clean_modified(self):
+        for _, node in self._nodes.items():
+            node.modified = False
+
+    def _closure_affected(self):
+        closure = set()
+        current = [id_ for id_, node in self._nodes.items() if node.modified]
+        # closure.update(current)
+        while current:
+            new_current = set()
+            for n in current:
+                new_neighs = self._inverse_neighbors(n)
+                to_add = set(new_neighs).difference(current)
+                new_current.update(to_add)
+                closure.update(to_add)
+            current = new_current
+
+        return closure
+
+    def _inverse_neighbors(self, node_id):
+        result = []
+        for id_, nodes_ids in self._edges.items():
+            if node_id in nodes_ids:
+                result.append(id_)
+        return result
+
+    def update_check_graph(self, deps_graph, output):
```

Comment:
Docstrings for every method and class in this module. In this case I don't understand the `check` name